### PR TITLE
Refactor issue retrieval methods

### PR DESCRIPTION
**Pull request by neithernut**

### CODE REVIEW

File: lib/src/issue.rs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        self.id
+    }
+
+    /// Get possible heads of the issue
+    ///
+    /// Returns the head references from both the local repository and remotes
+    /// for this issue.
+    ///
+    pub fn heads(&self) -> Result<References<'r>> {
+        let glob = format!("**/dit/{}/head", self);
```
Comment (matthiasbeyer) [inline lib/src/issue.rs]: Same as in the other issue: Don't use `self` here for formatting!

File: lib/src/issue.rs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            .chain_err(|| EK::WrappedGitError)
+    }
+
+    /// Get the local issue head for the issue
+    ///
+    /// Returns the head reference of the issue from the local repository, if
+    /// present.
+    ///
+    pub fn find_local_head(&self) -> Result<Reference<'r>> {
+        let refname = format!("refs/dit/{}/head", self);
```
Comment (matthiasbeyer) [inline lib/src/issue.rs]: Same here.

File: lib/src/issue.rs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            .chain_err(|| EK::WrappedGitError)
+    }
+
+    /// Get the leaf references for the issue
+    ///
+    /// Returns the leaf references for the issue from both the local repository
+    /// and remotes.
+    ///
+    pub fn issue_leaves(&self) -> Result<References<'r>> {
+        let glob = format!("**/dit/{}/leaves/*", self);
```
Comment (matthiasbeyer) [inline lib/src/issue.rs]: Same here.

File: lib/src/issue.rs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+            .chain_err(|| EK::WrappedGitError)
+    }
+
+    /// Get all local references for the issue
+    ///
+    /// Return all references associated with the issue from the local
+    /// repository.
+    ///
+    pub fn local_refs(&self) -> Result<References<'r>> {
+        let glob = format!("refs/dit/{}/**", self);
```
Comment (matthiasbeyer) [inline lib/src/issue.rs]: Same here.

File: lib/src/issue.rs [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            .references_glob(&glob)
+            .chain_err(|| EK::WrappedGitError)
+    }
+
+    /// Get a revwalk for traversing all messages of the issue
+    ///
+    /// The sorting of the revwalk will be set to "topological".
+    ///
+    pub fn message_revwalk(&self) -> Result<git2::Revwalk<'r>> {
+        let glob = format!("**/dit/{}/**", self);
```
Comment (matthiasbeyer) [inline lib/src/issue.rs]: Same here.

File: lib/src/repository.rs [hunk 6]
```diff
@@ -99,72 +86,56 @@ pub trait RepositoryExt {
...
-            .and_then(|mut revwalk| {
-                revwalk.push_glob(glob.as_ref())?;
-                revwalk.simplify_first_parent();
-                revwalk.set_sorting(git2::SORT_TOPOLOGICAL);
-                Ok(revwalk)
-            })
-            .chain_err(|| EK::WrappedGitError)
+    fn issue_by_head_ref(&self, head_ref: &git2::Reference) -> Result<Issue> {
+        head_ref.name()
+                .and_then(|name| if name.ends_with("/head") {Some(name)} else {None})
```
Comment (matthiasbeyer) [inline lib/src/repository.rs]: Don't do these if-else-oneliners. Nobody can read this without getting insane. Especially not without proper whitespace.

File: lib/src/repository.rs [hunk 7]
```diff
@@ -99,72 +86,56 @@ pub trait RepositoryExt {
...
         //       commit doesn't make sense for this function, semantically.
-        for c in FirstParentIter::new(commit.as_object().clone().into_commit().ok().unwrap()) {
-            let head = try!(self
-                            .get_issue_heads(c.id())
-                            .chain_err(|| EK::CannotFindIssueHead(c.id())));
-
-            if head.count() > 0 {
-                return Ok(c);
-            }
+        for c in FirstParentIter::new(message.as_object().clone().into_commit().ok().unwrap()) {
```
Comment (matthiasbeyer) [inline lib/src/repository.rs]: Any comment about why the `unwrap()` is safe would be nice.

File: lib/src/repository.rs [hunk 8]
```diff
@@ -99,72 +86,56 @@ pub trait RepositoryExt {
...
+            let issue = self.find_issue(c.id());
+            if issue.is_ok() { return issue }
```
Comment (matthiasbeyer) [inline lib/src/repository.rs]: And this is another oneliner which is ugly AF.

File: src/main.rs [hunk 9]
```diff
@@ -166,7 +166,7 @@ fn fetch_impl(repo: &Repository, matches: &clap::ArgMatches) -> i32 {
         // fetch a specific list of issues
         let iter = issues.map(Oid::from_str).abort_on_err();
         if matches.is_present("known") {
-            iter.chain(try_or_1!(repo.get_all_issue_hashes()).abort_on_err())
+            iter.chain(try_or_1!(repo.issues()).abort_on_err().map(|issue| issue.id()))
```
Comment (matthiasbeyer) [inline src/main.rs]: You could also try whether `map(Issue::id)` works here.

File: src/main.rs [hunk 10]
```diff
@@ -412,7 +414,7 @@ fn show_impl(repo: &Repository, matches: &clap::ArgMatches) -> i32 {
         if matches.is_present("initial") {
             vec![(TreeGraphElemLine::empty(), try_or_1!(repo.find_commit(issue)))]
         } else {
-            try_or_1!(repo.get_issue_revwalk(issue))
+            try_or_1!(repo.find_issue(issue).and_then(|issue| issue.message_revwalk()))
```
Comment (matthiasbeyer) [inline src/main.rs]: Also here: `and_then(Issue::message_revwalk)`?

File: src/main.rs [hunk 11]
```diff
@@ -463,7 +465,7 @@ fn tag_impl(repo: &Repository, matches: &clap::ArgMatches) -> i32 {
     let issue = try_or_1!(Oid::from_str(matches.value_of("issue-hash").unwrap()));
 
     // get the head for the issue to tag
-    let mut issue_head = try_or_1!(repo.get_local_issue_head(issue));
+    let mut issue_head = try_or_1!(repo.find_issue(issue).and_then(|issue| issue.find_local_head()));
```
Comment (matthiasbeyer) [inline src/main.rs]: Also here: `and_then(Issue::find_local_head)`?

File: lib/src/repository.rs (hunk 7 above)
Comment (neithernut) [inline lib/src/repository.rs]: Well, it's old code which will go away in the next round of refactors anyway.

File: src/main.rs (hunk 9 above)
Comment (neithernut) [inline src/main.rs]: Not directly, because `Issue::id` takes a reference.

File: src/main.rs (hunk 10 above)
Comment (neithernut) [inline src/main.rs]: Same as above.

File: src/main.rs (hunk 11 above)
Comment (neithernut) [inline src/main.rs]: Still same.

File: src/main.rs (hunk 9 above)
Comment (neithernut) [inline src/main.rs]: I think I tried.

File: lib/src/repository.rs [hunk 12]
```diff
@@ -92,29 +98,55 @@ impl RepositoryExt for Repository {
...
+                if name.ends_with("/head") {
+                    Some(name)
+                } else {
+                    None
+                }
+            })
+            .and_then(|name| name.rsplitn(3, "/").nth(1))
+            .ok_or_else(|| Error::from_kind(EK::MalFormedHeadReference(
+                name.unwrap_or_default().to_owned()
+            )))
```
Comment (matthiasbeyer) [inline lib/src/repository.rs]: Use a block `{...}` and variables here to make this readable.

File: lib/src/repository.rs [hunk 13]
```diff
@@ -92,29 +98,55 @@ impl RepositoryExt for Repository {
...
-    fn find_tree_init<'a>(&'a self, commit: &Commit<'a>) -> Result<Commit> {
+    fn issue_by_head_ref(&self, head_ref: &git2::Reference) -> Result<Issue> {
+        let name = head_ref.name();
+        name.and_then(|name| {
+                if name.ends_with("/head") {
+                    Some(name)
+                } else {
+                    None
+                }
+            })
```
Comment (matthiasbeyer) [inline lib/src/repository.rs]: You can remove one scope (around the if-else) here, IMO.

File: lib/src/repository.rs [hunk 14]
```diff
@@ -100,17 +100,18 @@ impl RepositoryExt for Repository {
...
         let name = head_ref.name();
-        name.and_then(|name| {
+        name.and_then(|name|
                 if name.ends_with("/head") {
                     Some(name)
                 } else {
                     None
                 }
-            })
+            )
```
Comment (matthiasbeyer) [inline lib/src/repository.rs]: I meant writing it this way:

```rust
name.and_then(|name| if name.ends_with("") {
 // ...
})
```

Comment (neithernut) [inline lib/src/repository.rs]: I absolutely can't tell on what basis you decide what's ugly and what's not.

Comment (neithernut) [inline lib/src/repository.rs]: I'm almost sure you did want me to change it the other way round at some point.

**Pull request merged**